src: kernel_config_manager: fix `kw (--save | -s)` behavior when git commit fails (CLONE)

davidbtadokoro opened the pull request:
After this patch, failing `kw (--save | -s)` because of error in commiting save operation of .config file no longer throws wrong error (undefined function) neither consolidates the operation.

Closes #696 

Obs.: This PR is a clone of PR #697. Check this thread for more info.

rodrigosiqueira commented on src/kernel_config_manager.sh:
```diff
@@ -140,6 +140,7 @@ function save_config_file()
   if [[ "$ret" == 1 ]]; then
     warning "Warning: $name: there's nothing new in this file"
   elif [[ "$ret" -gt 0 ]]; then
+    git reset --hard > /dev/null 2>&1 # Aborts save operation
```
Do we need `/dev/null`? Can we just add `--quiet`?

davidbtadokoro replied:
You are right 👍 
Doing it!

rodrigosiqueira commented on tests/kernel_config_manager_test.sh:
```diff
@@ -84,6 +84,8 @@ function test_save_config_file_check_save_failures()
 {
   local current_path="$PWD"
   local ret=0
+  local git_name=""
+  local git_email=""
```
Use single-quotes.

rodrigosiqueira commented on tests/kernel_config_manager_test.sh:
```diff
@@ -100,6 +102,19 @@ function test_save_config_file_check_save_failures()
+  touch .config
+  git_name=$(git config user.name)
+  git_email=$(git config user.email)
+  git config --global user.name ""
+  git config --global user.email ""
```
single quotes.

rodrigosiqueira commented on tests/kernel_config_manager_test.sh:
```diff
@@ -100,6 +102,19 @@ function test_save_config_file_check_save_failures()
+  git_email=$(git config user.email)
+  git config --global user.name ""
+  git config --global user.email ""
+  ret=$(save_config_file "$NO_FORCE $NAME_1" "$DESCRIPTION_1")
```
Add newline after ret

kwy95 replied:
Some of these tests should be on the previous commit, and as previously stated this commit seems unnecessary considering #723

kwy95 commented on src/kernel_config_manager.sh:
```diff
@@ -140,6 +140,7 @@ function save_config_file()
   if [[ "$ret" == 1 ]]; then
     warning "Warning: $name: there's nothing new in this file"
   elif [[ "$ret" -gt 0 ]]; then
+    git reset --hard --quiet # Aborts save operation
```
Given #723 I don't think that this commit is necessary.

davidbtadokoro replied:
Thanks for the clarification @kwy95, lesson learned 👍
Also think the commit is outdated with #723

kwy95 reviewed (changes requested):
Please resist the urge to close and reopen PRs with the same content, the discussions and reviews present in a PR are part of the history of the development process and you should always try to keep it coherent. To do this you should learn to manipulate the branch and commit you are working on to keep a clean set of updated commits based on the reviews and the upstream state.

rodrigosiqueira commented:
Hi @davidbtadokoro ,
Could you drop the second commit?
Also, could you include a test that covers the change in commit 1?

rodrigosiqueira commented:
Hey @davidbtadokoro , could you update this PR?

rodrigosiqueira commented:
Hey @davidbtadokoro , for this PR you just need to drop the second commit and add test for the first one.
Thanks

davidbtadokoro commented:
Hi @rodrigosiqueira , because of #723 isn't this PR obsolete at this point?

rodrigosiqueira commented:
> Hi @rodrigosiqueira , because of #723 isn't this PR obsolete at this point?

The first commit still valid.

Commit:
src: kernel_config_manager: replace undefined function call

In the function save_config_file(), there was an undefined function call
to the fail() function (part of the Shunit2 framework). This bug was
probably due to a confusion beetween the fail() and the complain()
functions. By replacing it with the complain() call, the bug was fixed.

Closes #696

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

davidbtadokoro commented:
Hi @rodrigosiqueira
 
As you requested, I dropped the second commit and actually redone the first because the commit message was outdated. However, I could not make the bug reproducible for testing, because #723 made the check dependent of a `cp` command. I.e. the only way to reproduce it was to find a deterministic way for the `cp` command to fail outside the function and I couldn't find a way to do it.

rodrigosiqueira commented:
Excellent, now we just need a test to cover that line. Could you add it?

davidbtadokoro commented:
Hi @rodrigosiqueira 

I am sorry, but I could not figure out a way to reproduce the bug for testing, i.e. make the `cp` command to fail. Please, see last comment I made.

rodrigosiqueira commented:
Merged, thanks
